psi4/src/psi4/libmints/mintshelper.cc

```diff
@@ -1215,57 +1163,57 @@ SharedMatrix MintsHelper::mo_spin_eri_helper(SharedMatrix Iso, int n1, int n2) {
     return Ispin;
 }
 
-SharedMatrix MintsHelper::so_overlap() {
+bool MintsHelper::are_ints_cached(const std::string &label) {
+    auto it = cached_oe_ints_.find(label);
+    return it != cached_oe_ints_.end();
+}
+
+void MintsHelper::cache_ao_to_so_ints(SharedMatrix ao_ints, const std::string &label) {
     if (factory_->nirrep() == 1) {
-        SharedMatrix ret = ao_overlap();
-        ret->set_name(PSIF_SO_S);
-        return ret;
+        //        ao_ints.set_name(label);
+        cached_oe_ints_[label] = ao_ints;
     } else {
-        SharedMatrix overlap_mat(factory_->create_matrix(PSIF_SO_S));
-        overlap_mat->apply_symmetry(ao_overlap(), petite_list()->aotoso());
-        return overlap_mat;
+        SharedMatrix so_ints_sym(factory_->create_matrix(label));
+        so_ints_sym->apply_symmetry(ao_ints, petite_list()->aotoso());
+        cached_oe_ints_[label] = so_ints_sym;
     }
 }
 
-SharedMatrix MintsHelper::so_kinetic() {
+SharedMatrix MintsHelper::so_overlap_nr() {
```

Comment:
Should the `_nr` functions be options to the non-`nr`functions?

```
SharedMatrix MintsHelper::so_kinetic(std::string relativistic = "auto")
```

Reply:
I found it hard to come up with a good solution here. The problem is that in X2C one has to compute S, T, and V all at once and this computation needs the non-relativistic S, T, and V.  I used this somewhat redundant form to clearly separate the function that handles both relativistic and non-relativist cases from the actual computation of non-relativistic overlap. I can adopt the solution that you suggest, but that will introduce a recursive calling to the `so_kinetic` function. If that's not considered too bad then I can give it a try. Let me know.